spec/mem/tree_borrows/memory.md

```diff
@@ -138,18 +141,30 @@ impl<T: Target> TreeBorrowsMemory<T> {
     fn release_protector(&mut self, provenance: TreeBorrowsProvenance) -> Result {
         let (alloc_id, path) = provenance;
         self.mem.allocations.mutate_at(alloc_id.0, |allocation| {
-            let protected_node = allocation.extra.root.get_node(path);
-            assert_ne!(protected_node.protected, Protected::No);
+            let protector_accesses : List<_> = allocation.extra.root.access_node(path, |node| {
+                assert!(node.protected.yes());
+                // We get a list of pairs, and need to turn this into a pair of lists.
+                let mut new_perms = list![];
+                let mut derived_accs = list![];
+                for permission in node.permissions {
+                    // One list contains the first component, the other the second.
+                    let (new_perm, derived_access) = permission.unprotect();
+                    new_perms.push(new_perm);
+                    derived_accs.push(derived_access);
+                }
```

Comment:
can we turn this into building an intermediate list of pairs and then mapping both `fst` and `snd` onto it to create the two new lists?